File: Data Structures & Algorithms/same-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (!p && !q) return true;

        stack<TreeNode*> pst, qst;
        pst.push(p);
        qst.push(q);

        TreeNode *pit, *qit;
        while (!pst.empty() || !qst.empty()) {
            pit = pst.top(); pst.pop();
            qit = qst.top(); qst.pop();

            if ((pit && !qit) || (!pit && qit)) return false;
            if (pit && qit && pit->val != qit->val) return false;

            if (pit->left && qit->left) {
                pst.push(pit->left);
                qst.push(qit->left);
            } else if (pit->left || qit->left) return false;


            if (pit->right && qit->right) {
                pst.push(pit->right);
                qst.push(qit->right);
            } else if (pit->right || qit->right) return false;
        }
        return true;
    }
};

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []

        ans = []
        q = []
        q.append(root)

        while q:
            s = len(q)
            ans.append([n.val for n in q])
            for _ in range(s):
                curr = q[0]
                q.pop(0)
                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)
        return ans